virttest/qemu_devices/qcontainer.py

```diff
@@ -1483,51 +1515,105 @@ def define_hbas(qtype, atype, bus, unit, port, qbus, pci_bus,
+            devices[-1].set_param('boot', boot, bool)
+            devices[-1].set_param('snapshot', snapshot, bool)
+            devices[-1].set_param('readonly', readonly, bool)
+        else:
+            if snapshot is not None:
+                logging.warn('Snapshot is not supported in blockdev.')
+
         if 'aio' in self.get_help_text():
             if aio == 'native' and snapshot == 'yes':
                 logging.warn('snapshot is on, fallback aio to threads.')
                 aio = 'threads'
-            devices[-1].set_param('aio', aio)
+            if not self.caps['support_blockdev']:
+                devices[-1].set_param('aio', aio)
+            else:
+                if protocol_type in ['file', 'host_device', 'host_cdrom']:
+                    devices[-2].set_param('aio', aio)
             if aio == 'native':
                 # Since qemu 2.6, aio=native has no effect without
                 # cache.direct=on or cache=none, It will be error out.
                 # Please refer to qemu commit d657c0c.
                 cache = cache not in ['none', 'directsync'] and 'none' or cache
-        # Forbid to specify the cache mode for empty drives.
-        # More info from qemu commit 91a097e74.
-        if not filename:
-            cache = None
-        devices[-1].set_param('cache', cache)
-        devices[-1].set_param('media', media)
-        devices[-1].set_param('format', imgfmt)
-        if blkdebug is not None:
-            devices[-1].set_param('file', 'blkdebug:%s:%s' % (blkdebug,
-                                                              filename))
+        if not self.caps['support_blockdev']:
+            if blkdebug is not None:
+                devices[-1].set_param('file', 'blkdebug:%s:%s' % (blkdebug,
+                                                                  filename))
+            else:
+                devices[-1].set_param('file', filename)
+            devices[-1].set_param('cache', cache)
+            devices[-1].set_param('media', media)
+            devices[-1].set_param('format', imgfmt)
         else:
-            devices[-1].set_param('file', filename)
+            devices[-2].set_param('filename', filename)
+            for dev in (devices[-1], devices[-2]):
+                dev.set_param('cache.direct',
+                              self.cache_map[cache]['cache.direct'])
+                dev.set_param('cache.no-flush',
+                              self.cache_map[cache]['cache.no-flush'])
+            if media == 'cdrom':
+                imgfmt = 'raw'
+                readonly = 'on'
+            devices[-2].set_param('read-only', readonly, bool)
+            devices[-1].set_param('read-only', readonly, bool)
+            devices[-1].set_param('driver', imgfmt)
+            devices[-2].set_param('driver', protocol_type)
+            # Bounding the format node and protocol node with "node-name".
+            devices[-1].set_param('file', devices[-2].get_qid())
```

Comment:
It should be done by the node obj itself when you specify the parent-child relationship.